### PR TITLE
Add support for extracting target images from fat offload binaries

### PR DESCRIPTION
**Pull request by sndmitriev**
This patch consists of two parts

1) Change clang-offload-wrapper to create <address, size> pairs for all target
   images and add them to a .tgtimg section in the wrapper object. Contents of
   this section can later be used for locating and extracting target images
   from the fat offload binary using the new clang-offload-extract tool.

2) Add clang-offload-extract tool which allows extracting embedded target
   images using information from the .tgtimg section.

Signed-off-by: Sergey Dmitriev <serguei.n.dmitriev@intel.com>

### CODE REVIEW

**Review by s-kanaev — DISMISSED** [review]
SYCL RT CmakeLists files changes look good.

File: clang/test/Driver/clang-offload-extract.c
Comment (kbobrovs) [inline clang/test/Driver/clang-offload-extract.c:25]: please add a test case with -kind=sycl

**Review by kbobrovs — COMMENTED** [review]
Except for the test case request - LGTM (as we already did internal review).
That is a really useful tool, many people asked for this functionality.

adding @pvchupin

**Review by kbobrovs — DISMISSED** [review]

File: clang/tools/clang-offload-extract/ClangOffloadExtract.cpp
Comment (mlychkov) [inline clang/tools/clang-offload-extract/ClangOffloadExtract.cpp:29]:
```suggestion
#define IMAGE_INFO_SECTION_NAME ".tgtimg"
```

Comment (sndmitriev) [reply clang/tools/clang-offload-extract/ClangOffloadExtract.cpp:29]: Thanks! I did not notice that I have made a typo here.

File: clang/tools/clang-offload-extract/ClangOffloadExtract.cpp
Comment (sndmitriev) [inline clang/tools/clang-offload-extract/ClangOffloadExtract.cpp:104]:
```suggestion
    if (*NameOrErr != IMAGE_INFO_SECTION_NAME)
```

**Review by mdtoguchi — APPROVED** [review]
LGTM

**Review by kbobrovs — APPROVED** [review]

**Review by s-kanaev — APPROVED** [review]
SYCL RT CMakeLists.txt changes LGTM

**Review by AGindinson — DISMISSED** [review]

**Comment by bader** [comment]
@sndmitriev, should we add a separate code owners for this tool?
I think, front-end compiler owners will be assigned by default.
https://github.com/intel/llvm/blob/sycl/.github/CODEOWNERS#L3-L4

**Comment by sndmitriev** [comment]
> @sndmitriev, should we add a separate code owners for this tool?
> I think, front-end compiler owners will be assigned by default.
> https://github.com/intel/llvm/blob/sycl/.github/CODEOWNERS#L3-L4

Front-end is probably not the right component for it. I will add a new line for it to CODEOWNERS.

**Comment by sndmitriev** [comment]
Does this patch need any additional changes to be merged?

**Review by kbobrovs — APPROVED** [review]

**Comment by bader** [comment]
> Does this patch need any additional changes to be merged?

I don't think so. It's waiting on code owner review from @AaronBallman, @elizabethandrews, and/or @premanandrao, but I think cmake changes in clang project looks okay, so I'm going to merge this patch. Fee free to leave your comments anyway.

**Comment by kbobrovs** [comment]
@AaronBallman @elizabethandrews  @premanandrao, please review few changes in FE area.

**Comment by kbobrovs** [comment]
> @AaronBallman @elizabethandrews @premanandrao, please review few changes in FE area.

Oh, it has been merged already - please disregard.

**Pull request merged**
